### PR TITLE
Fixes to some issues reported by static analyzer in tracking related code

### PR DESCRIPTION
**Pull request by makortel**
This PR attempts to fix some of the issues reported by static analyzer in tracking (and related) code.

Tested in CMSSW_7_6_X_2015-06-30-2300, no changes expected in standard workflows.

@rovere @VinInn

### CODE REVIEW

**Comment by diguida** [comment]
please test

**Comment by deguio** [comment]
+1

**Comment by mmusich** [comment]
+1 
minor changes in BSFitter code and TrackerHitFilter affecting AlCa

**Comment by slava77** [comment]
@cmsbuild please test
I don't see this test in jenkins lineup

File: RecoVertex/VertexTools/src/DeterministicAnnealing.cc [hunk 1]
```diff
@@ -36,7 +37,7 @@ double DeterministicAnnealing::weight ( double chi2 ) const
   */
   // return mphi / ( mphi + phi ( theChi2cut ) );
   long double newtmp = mphi / ( mphi + phi ( theChi2cut ) );
```
Comment (VinInn) [inline RecoVertex/VertexTools/src/DeterministicAnnealing.cc]: long double???
this is a joke  (next round we should use "double")

**Comment by slava77** [comment]
I suppose "some " in the subject of this PR is the key ;)
there are more things reported even in the files touched here
https://slava77sk.web.cern.ch/slava77sk/reco/llvm/10066/
I suggest further changes are done in the next PR

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/10066	5f95bc2
- since jenkins is stuck, tested in CMSSW_7_6_X_2015-07-04-2300 /test area sign570/
- the static analyzer warnings were cleaned up considerably https://slava77sk.web.cern.ch/slava77sk/reco/llvm/10066/
- short matrix tests pass and comparisons with baseline show no difference

**Comment by makortel** [comment]
Of course clang wasn't happy about that... I can take a look on it next week.

**Comment by VinInn** [comment]
I have a final solution
https://github.com/VinInn/cmssw/blob/6127ed0ab28919992e4c072bde3cebd70b1319fc/CommonTools/Utils/interface/DynArray.h
Maybe I will make a PR with only that...

**Comment by makortel** [comment]
Thanks, that should do it (I was also thinking that I saw something related in some your PRs). Please go ahead if you want, I can then adjust this one as needed.

**Comment by makortel** [comment]
I removed the commit causing problems for clang (\*), switched string comparisons to (const) strongly-typed enum in SiStripTrackingRecHitsValid (**), and rebased on top of CMSSW_7_6_X_2015-07-13-1100.

(\*) while waiting for Vincenzo's DynArray
(**) in the hope that now static analyzer would not assume that the content of `matchedmonorstereo` could change between the comparisons (I say "hope" because for me `scram b checker` crashes for most of the files, including SiStripTrackingRecHitsValid.cc)

**Comment by deguio** [comment]
+1

**Comment by slava77** [comment]
@cmsbuild please test
IBs are OK now

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/10066	2738698
- compared to the previous "+1" for 5f95bc2, changes are only to back-out a change (revert to baseline) in HitPairGeneratorFromLayerPair.cc [related to reco code] and in SiStripTrackingRecHitsValid.h [only in validation]
- jenkins tests passed and comparisons with baseline show no (relevant) differrence

**Pull request merged**